keps/1282-requeue-strategy/README.md

```diff
@@ -0,0 +1,306 @@
+
+	if p1 != p2 {
+		return p1 > p2
+	}
+
+	tA := workload.GetQueueOrderTimestamp(objA.Obj)
+	tB := workload.GetQueueOrderTimestamp(objB.Obj)
+	return !tB.Before(tA)
+}
+```
+
+```go
+package workload
+
+// ...
+
+func GetQueueOrderTimestamp(w *kueue.Workload) *metav1.Time {
+	if c := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadEvicted); c != nil && c.Status == metav1.ConditionTrue && c.Reason == kueue.WorkloadEvictedByPodsReadyTimeout {
+		return &c.LastTransitionTime
+	}
+	return &w.CreationTimestamp
+}
+```
+
+#### Proposed Sorting
+
+Add types that contain information about queue ordering strategy to the `apis/config/<version>` package. For example:
+
+```go
+package v1beta1
+
+// ...
+
+type QueueOrderingCriteria string
+
+const (
+  OrderOnEvictionTimestamp = QueueOrderingCriteria("UseEvictionTimestamp")
+  OrderOnCreationTimestamp = QueueOrderingCriteria("UseCreationTimestamp")
+)
+
+type QueueOrderingStrategy struct{
+  RequeueOnPriorityPreemption QueueOrderingCriteria
+  RequeueOnPodsReadyTimeout   QueueOrderingCriteria
+}
+```
+
+The `pkg/workload` package could have the `GetQueueOrderTimestamp` function modified to include conditionals that control which timestamp to return based on the configured ordering strategy.
+
+```go
+func GetQueueOrderTimestamp(w *kueue.Workload, s QueueOrderingStrategy) *metav1.Time {
+  switch getEvictionReason(w) {
+    case kueue.WorkloadEvictedByPodsReadyTimeout:
+      if s.RequeueOnPodsReadyTimeout == OrderOnEvictionTimestamp {
+        return &c.LastTransitionTime
+      } else {
+        return &c.CreationTimestamp
+      }
+    case kueue.WorkloadEvictedByPreemption:
+      if s.RequeueOnPriorityPreemption == OrderOnEvictionTimestamp {
+        return &c.LastTransitionTime
+      } else {
+        return &c.CreationTimestamp
+      }
+    default:
+      return &w.CreationTimestamp
+  }
+}
+
+// ...
+
+func getEvictionReason(w *kueue.Workload) string {
+	if c := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadEvicted); c != nil && c.Status == metav1.ConditionTrue {
+    return c.Reason
+  }
+  return ""
+}
+```
+
+#### Preemption/Preemptor Flapping
+
+To address the [following concern](https://github.com/kubernetes-sigs/kueue/issues/1282#issue-1966106166):
+
+> The case of preemption might be more tricky: we want the preempted jobs to be readmitted as soon as possible. But if a job is waiting for more than one job to be preempted and the queueing strategy is BestEffortFIFO, we don't want the preempted pods to take the head of the queue.
+> Maybe we need to hold them until the preemptor job is admitted, and then they should use the regular priority.
+
+The `pkg/queue` package could have the existing `queueOrderingFunc()` modified to add sorting based on who is the preemptor (might need to add a condition to the Workload to track this).
+
+
+```go
+package queue 
+
+// ...
+
+func queueOrderingFunc(s kueue.QueueOrderingStrategy) func(interface{}, interface{}) bool {
```

Comment:
Similarly here, I don't see a need for the code. We typically just record the API in the `Design Details` section, and describe the mechanism / handling of the API.

Reply:
Thanks, I can remove this.